tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java

```diff
@@ -0,0 +1,355 @@
+                    List<String> topicsToCreate = new ArrayList<>();
+                    for (int i = 0; i < spec.totalTopics(); i++) {
+                        topicsToCreate.add(topicIndexToName(i));
+                    }
+                    log.info("Creating " + spec.totalTopics() + " topics...");
+                    List<Future<Void>> futures = new ArrayList<>();
+                    while (!topicsToCreate.isEmpty()) {
+                        List<NewTopic> newTopics = new ArrayList<>();
+                        for (int i = 0; (i < MAX_BATCH_SIZE) && !topicsToCreate.isEmpty(); i++) {
+                            String topic = topicsToCreate.remove(0);
+                            newTopics.add(new NewTopic(topic, 1, REPLICATION_FACTOR));
+                        }
+                        futures.add(adminClient.createTopics(newTopics).all());
+                    }
+                    for (Future<Void> future : futures) {
+                        future.get();
+                    }
+                    log.info("Successfully created " + spec.totalTopics() + " topics.");
+                }
+                executor.submit(new SendRecords());
+            } catch (Exception e) {
+                abort("CreateBenchmarkTopics", e);
+            }
+            return null;
+        }
+    }
+
+    private static class SendRecordsCallback implements Callback {
+        private final SendRecords sendRecords;
+        private final long startMs;
+
+        SendRecordsCallback(SendRecords sendRecords, long startMs) {
+            this.sendRecords = sendRecords;
+            this.startMs = startMs;
+        }
+
+        @Override
+        public void onCompletion(RecordMetadata metadata, Exception exception) {
+            long now = Time.SYSTEM.milliseconds();
+            long durationMs = now - startMs;
+            sendRecords.recordDuration(durationMs);
+            if (exception != null) {
+                log.error("SendRecordsCallback: error", exception);
+            }
+        }
+    }
+
+    private int perSecToPerPeriod(float perSec, long periodMs) {
+        float period = ((float) periodMs) / 1000.0f;
+        float perPeriod = perSec * period;
+        perPeriod = Math.max(1.0f, perPeriod);
+        return (int) perPeriod;
+    }
+
+    /**
+     * A subclass of Throttle which flushes the Producer right before the throttle injects a delay.
+     * This avoids including throttling latency in latency measurements.
+     */
+    private static class SendRecordsThrottle extends Throttle {
+        private final KafkaProducer<?, ?> producer;
+
+        SendRecordsThrottle(int maxPerPeriod, KafkaProducer<?, ?> producer) {
+            super(maxPerPeriod, THROTTLE_PERIOD_MS);
+            this.producer = producer;
+        }
+
+        @Override
+        protected synchronized void delay(long amount) throws InterruptedException {
+            producer.flush();
```

Comment:
@cmccabe Since `flush` blocks until all records are sent, wouldn't it be better to compute the delay time after flush completes? Ideally, an async flush would be better to avoid more delay than required, but that would need another thread. Alternative is not to flush at all, since only the records in the last incomplete batch would be delayed when `linger.ms > 0`.

Reply:
Good point.  I updated this so that it checks the time before and after, and only waits for the remaining time.  I think the flush is still useful because we might want to test with linger.ms > 0 (as many users have configured it).